File: src/pl/swmud/ns/swaedit/gui/ExitEventFilter.java
package pl.swmud.ns.swaedit.gui;

import com.trolltech.qt.core.QEvent;
import com.trolltech.qt.core.QObject;
import com.trolltech.qt.gui.QMouseEvent;
import com.trolltech.qt.gui.QToolButton;

class ExitEventFilter extends QObject {

    public boolean eventFilter(QObject o, QEvent e) {
        QToolButton button = (QToolButton)o; 
        if (e.type() == QEvent.Type.MouseButtonPress && button.isEnabled()) {
            button.pressed.emit();
            return true;
        }
        if (e instanceof QMouseEvent) {
            return true;
        }
        return super.eventFilter(o, e);
    }

}

File: src/gui/ExitEventFilter.cpp
#include "gui/ExitEventFilter.h"

#include <QEvent>
#include <QMouseEvent>

ExitEventFilter::ExitEventFilter(QObject *parent)
    : QObject(parent)
{
}

bool ExitEventFilter::eventFilter(QObject *watched, QEvent *event)
{
    if (event->type() == QEvent::MouseButtonPress) {
        QMouseEvent *me = static_cast<QMouseEvent *>(event);
        Q_UNUSED(me);
        emit pressed();
        return true;
    }

    // Swallow all other mouse events
    if (event->type() == QEvent::MouseButtonRelease ||
        event->type() == QEvent::MouseButtonDblClick ||
        event->type() == QEvent::MouseMove) {
        return true;
    }

    return QObject::eventFilter(watched, event);
}
